[Inductor] Pattern matcher support for mutable ops with non-view inputs

yf225 opened the pull request:
Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #152776
* __->__ #152775

Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

yf225 committed:
[Inductor] Pattern matcher support for mutable ops with non-view inputs

ghstack-source-id: daace000b4fa40bdc2727aefabb354664cea9540
Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -1655,6 +1657,78 @@ def my_func_static(x, w, epsilon):
+        def bar_out(x: torch.Tensor, out: torch.Tensor) -> None:
+            out.copy_(x + 2)
+
+        @register_fake("mylib::bar")
+        def bar_out_fake(x: torch.Tensor, out: torch.Tensor) -> None:
+            return None
```
Don't need a register_fake if the custom op returns None

zou3519 commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -1655,6 +1657,78 @@ def my_func_static(x, w, epsilon):
+            search_fn=mutable_ops_pattern,
+            replace_fn=mutable_ops_replacement,
+            example_inputs=[inp.clone().detach(), inp.clone().detach()],
+            trace_fn=functools.partial(fwd_only, apply_auto_functionalize=True),
```
1) Should apply_auto_functionalize be the default for fwd_only? Since Inductor pattern matching happens only at the post-grad pass?
2) If not, should we rename `apply_auto_functionalize` something like `post_grad`? Because that is where we plan to use the pattern matcher

zou3519 commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -1655,6 +1657,78 @@ def my_func_static(x, w, epsilon):
+        register_replacement(
+            search_fn=mutable_ops_pattern,
+            replace_fn=mutable_ops_replacement,
+            example_inputs=[inp.clone().detach(), inp.clone().detach()],
```
We are technically exposing fwd_only as public API.

Instead of doing that I think it might be better for PatternMatcherPass to get an enum about the location it's operating on (e.g. post-grad functional IR).

zou3519 commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -1655,6 +1657,78 @@ def my_func_static(x, w, epsilon):
+        # user-function
+        @torch.compile(fullgraph=True, backend=custom_backend)
+        def f(x):
+            x = x.clone()
+            out = torch.zeros_like(x)
+            foo_inplace(x)
+            bar_out(x, out)
+            return out
```
Could you add a test removing the x.clone() (and therefore make it so that we're mutating the input) and check that the resulting compiled function (after the pattern matching) still mutates the inputs (or fails to pattern match?)

yf225 replied:
Added `# Case 2: mutates graph input` below - it seems to still work

yf225 replied:
Added `# Case 2: mutates graph input` case below - it seems to fail to pattern match due to the extra copy_ op at end of the graph.

yf225 replied:
I believe it's because `copy_` uses `getitem_1`, hence `getitem_1` has 2 users instead of 1 (expected by the pattern), thus causing the mismatch due to this line: https://github.com/pytorch/pytorch/blob/7cf8049d63ee7e8632f2e7332a654d4f2b9550a4/torch/_inductor/pattern_matcher.py#L595.

yf225 replied:
Perhaps we can introduce a new `torch._inductor.graph_input(...)` fake op (in the same spirit as opaque_view), and codegen the copy_ op when tracing through it + afv2?

zou3519 replied:
graph_input seems reasonable then, I don't see how else we are going to do this

eellison replied:
Can we discuss this a bit more? i'm worried if we have to add graph_input this is going to be too unergonomic/surprising to use.

yf225 replied:
I got the example working with Richard's suggestion of returning all mutated inputs in the pattern. But I think the # users matching problem still stands in general: https://github.com/pytorch/pytorch/issues/153208.

yf225 replied:
I got the example working with Richard's suggestion of returning all mutated inputs in the pattern. So we don't need the graph_input new API.

zou3519 reviewed (commented):
This looks good to me. I suggested another test case, plus we should bikeshed the API since this should be "public".

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -2082,6 +2083,7 @@ def fwd_only(
     fn: Callable[..., Any],
     args: Sequence[Any],
     *,
+    apply_auto_functionalize: bool = False,
```
+1 


```
    apply_auto_functionalize: bool = False,
    run_functional_passes: bool = True,
```

How do these two options relate to each other ? when would you do one but not the other ?

yf225 replied:
discussed with Richard, currently:
- `run_functional_passes=True` means "run DCE on the pattern and replacement graphs" (which is not safe to use for reinplace pass), maybe we can rename it to `run_dce_and_noop_elimination`.
- `apply_auto_functionalize=True` means "run functionalize on both pattern and replacement graphs", maybe we can rename it to `functionalize_patterns`.

Q: Can we merge the two flags?

Seems in general they are the same value:

- if "non_functional_ir", both False
- if "functional_ir", both True
- However pre_grad FX IR is an exception right now: it's "non_functional_ir", but it currently has `run_functional_passes=True`; so we might need to change it to False, if we want to merge the two flags.

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

yf225 committed:
Update on "[Inductor] Pattern matcher support for mutable ops with non-view inputs"

Fixes the non-view input use case in https://github.com/pytorch/pytorch/issues/152441.


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/152767

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
